modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,852 @@
+  for zippered iteration). Adding leader/follower support is a future goal.
+
+  This module also exposes some helper procedures used to break files into chunks.
+  These could be used as building blocks to implement other parallel I/O routines:
+
+  * :proc:`findDelimChunks`: find a set of byte offsets that divide a file into
+    roughly equal chunks where each chunk begins with a delimiter
+  * :proc:`findItemOffsets`: get a prefix sum of the number of items in each
+    chunk of a file, where the chunks are defined by the ``byteOffsets`` array,
+    and each item is strictly separated by the given delimiter
+  * :proc:`findDelimChunksChecked`: find a set of byte offsets that divide a file
+    into roughly equal chunks where each chunk begins with a delimiter and each
+    chunk starts with a deserializable value of the given type
+*/
+@unstable("the 'parallelIO' module is unstable and subject to change in a future release")
+module ParallelIO {
+  private use IO, BlockDist, List, CTypes, OS;
+
+  /*
+    Iterate over a file's lines in parallel.
+
+    This routine is similar to :proc:`readLinesAsArray`, except that it yields
+    each line as it is read instead of returning an array of lines.
+
+    .. warning::
+
+      This routine will halt if the file cannot be opened or if an
+      I/O error occurs while reading the file. This limitation is
+      expected to be removed in a future release.
+
+    .. note::
+
+      Only serial and standalone-parallel iteration is supported. This
+      iterator cannot yet be used in a zippered context.
+
+    :arg filePath: a path to the file to read from
+    :arg lineType: which type to return for each line — either ``string`` or ``bytes``
+    :arg header: how to handle the file header (see :record:`headerPolicy`)
+    :arg nTasks: the number of tasks used to read the file
+  */
+  iter readLines(filePath: string, type lineType = string,
+                 header = headerPolicy.noHeader, nTasks: int = here.maxTaskPar
+  ): lineType
+    where lineType == string || lineType == bytes
+  {
+    param delim = b"\n";
+
+    const f = try! open(filePath, ioMode.r),
+          (findStart, startOffset) = try! header.apply(f, 0..<f.size),
+          r = try! f.reader(locking=false, region=startOffset..);
+
+    if findStart then try! r.advanceThrough(delim);
+
+    var line: lineType;
+    while (try! r.readLine(line, stripNewline=true)) do
+      yield line;
+  }
+
+  @chpldoc.nodoc
+  iter readLines(param tag: iterKind, filePath: string, type lineType = string,
+                 header = headerPolicy.noHeader, nTasks: int = here.maxTaskPar
+  ): lineType
+    where tag == iterKind.standalone && (lineType == string || lineType == bytes)
+  {
+    param delim = b"\n";
+
+    const fMeta = try! open(filePath, ioMode.r),
+          fileBounds = 0..<(try! fMeta.size),
+          byteOffsets = try! findDelimChunks(fMeta, delim, nTasks, fileBounds, header);
```

Comment:
Thanks for adding a standalone parallel `readLines`. This doesn't work the way I was envisioning in https://github.com/chapel-lang/chapel/pull/23876#discussion_r1447629039 . Could you add updating it to be 1-pass in that way as Future Work?

Reply:
Right, I think it definitely could be single pass. I just went with this implementation for now because it was expedient and because the leader-iterator version will need to compute `byteOffets` and `itemOffsets` to communicate those with followers anyway (I think).